### PR TITLE
Move math types over to nalgebra with Generic dimensions > 2 and f32/f64 support

### PR DESCRIPTION
**Pull request by marstaik**
BVH should now work (in-theory) with 2d,3d,4d+ and with f32,f64 using the base nalgebra types.

Consistent performance, everything does mostly get SIMD. On my older first gen threadripper, I see up to 10% performance improvements across the board. Those that match or are slightly above are usually within error.

Axis and some re-exports were simply deleted as they aren't needed when using the nalgebra types.

The only outliers are:
* `bench_intersects_aabb`, which runs a bit slower than the fastest variant of before nalgebra
    * I wrote this to use pure SIMD, but it doesn't optimize as well as it should. I should be able to make this much faster with some manual SIMD later.
* `bench_optimize_bvh_*` The optimization tests run over 200% faster, and blow the old optimize code out the water. It is actually faster to optimize now than to rebuild.

| test| glam | nalgebra |
|----|------|------|
| test bvh::bvh_impl::bench::bench_build_1200_triangles_bvh                |     821,735 ns/iter (+/- 67,897)       |     771,730 ns/iter (+/- 21,811) |
| test bvh::bvh_impl::bench::bench_build_120k_triangles_bvh                | 103,104,360 ns/iter (+/- 4,907,146)    | 101,412,600 ns/iter (+/- 4,414,230) |
| test bvh::bvh_impl::bench::bench_build_12k_triangles_bvh                 |   8,944,690 ns/iter (+/- 267,942)      |   8,685,235 ns/iter (+/- 131,969) |
| test bvh::bvh_impl::bench::bench_build_sponza_bvh                        |  79,784,630 ns/iter (+/- 2,508,824)    |  78,247,340 ns/iter (+/- 1,407,163) |
| test bvh::bvh_impl::bench::bench_intersect_1200_triangles_bvh            |         130 ns/iter (+/- 2)            |         149 ns/iter (+/- 6) |
| test bvh::bvh_impl::bench::bench_intersect_120k_triangles_bvh            |         765 ns/iter (+/- 14)           |         863 ns/iter (+/- 12) |
| test bvh::bvh_impl::bench::bench_intersect_12k_triangles_bvh             |         321 ns/iter (+/- 6)            |         371 ns/iter (+/- 5) |
| test bvh::bvh_impl::bench::bench_intersect_sponza_bvh                    |       2,289 ns/iter (+/- 64)           |       1,905 ns/iter (+/- 36) |
| test bvh::iter::bench::bench_intersect_128rays_sponza_iter               |     159,109 ns/iter (+/- 4,056)        |     167,965 ns/iter (+/- 3,013) |
| test bvh::iter::bench::bench_intersect_128rays_sponza_vec                |     293,305 ns/iter (+/- 7,316)        |     244,404 ns/iter (+/- 4,788) |
| test bvh::optimization::bench::bench_intersect_120k_after_optimize_00p   |         764 ns/iter (+/- 12)           |         864 ns/iter (+/- 10) |
| test bvh::optimization::bench::bench_intersect_120k_after_optimize_01p   |     128,246 ns/iter (+/- 10,879)       |     145,156 ns/iter (+/- 11,400) |
| test bvh::optimization::bench::bench_intersect_120k_after_optimize_10p   |   1,634,400 ns/iter (+/- 387,476)      |   1,761,105 ns/iter (+/- 425,368) |
| test bvh::optimization::bench::bench_intersect_120k_after_optimize_50p   |   2,234,812 ns/iter (+/- 440,313)      |   2,395,035 ns/iter (+/- 498,051) |
| test bvh::optimization::bench::bench_intersect_120k_with_rebuild_00p     |         764 ns/iter (+/- 19)           |         862 ns/iter (+/- 7) |
| test bvh::optimization::bench::bench_intersect_120k_with_rebuild_01p     |         824 ns/iter (+/- 29)           |         928 ns/iter (+/- 21) |
| test bvh::optimization::bench::bench_intersect_120k_with_rebuild_10p     |       1,843 ns/iter (+/- 62)           |       2,000 ns/iter (+/- 39) |
| test bvh::optimization::bench::bench_intersect_120k_with_rebuild_50p     |       2,128 ns/iter (+/- 46)           |       2,282 ns/iter (+/- 77) |
| test bvh::optimization::bench::bench_intersect_sponza_after_optimize_00p |       1,613 ns/iter (+/- 51)           |       1,392 ns/iter (+/- 25) |
| test bvh::optimization::bench::bench_intersect_sponza_after_optimize_01p |       2,495 ns/iter (+/- 113)          |       3,051 ns/iter (+/- 58) |
| test bvh::optimization::bench::bench_intersect_sponza_after_optimize_10p |       3,781 ns/iter (+/- 153)          |       4,421 ns/iter (+/- 180) |
| test bvh::optimization::bench::bench_intersect_sponza_after_optimize_50p |       5,534 ns/iter (+/- 394)          |       6,352 ns/iter (+/- 220) |
| test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_00p   |       1,609 ns/iter (+/- 36)           |       1,390 ns/iter (+/- 26) |
| test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_01p   |       1,729 ns/iter (+/- 36)           |       1,527 ns/iter (+/- 26) |
| test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_10p   |       2,074 ns/iter (+/- 144)          |       1,948 ns/iter (+/- 44) |
| test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_50p   |       2,665 ns/iter (+/- 71)           |       2,578 ns/iter (+/- 51) |
| test bvh::optimization::bench::bench_optimize_bvh_120k_00p               |   1,090,720 ns/iter (+/- 23,355)       |   1,164,270 ns/iter (+/- 79,435) |
| test bvh::optimization::bench::bench_optimize_bvh_120k_01p               |   7,373,650 ns/iter (+/- 1,720,855)    |   2,237,075 ns/iter (+/- 218,183) |
| test bvh::optimization::bench::bench_optimize_bvh_120k_10p               |  40,201,330 ns/iter (+/- 14,485,239)   |  12,243,750 ns/iter (+/- 1,361,338) |
| test bvh::optimization::bench::bench_optimize_bvh_120k_50p               | 222,987,100 ns/iter (+/- 68,341,041)   |  55,583,600 ns/iter (+/- 13,194,002) |
| test bvh::optimization::bench::bench_randomize_120k_50p                  |   6,262,345 ns/iter (+/- 360,555)      |   5,717,540 ns/iter (+/- 229,427) |
| test flat_bvh::bench::bench_build_1200_triangles_flat_bvh                |     676,387 ns/iter (+/- 16,147)       |     658,178 ns/iter (+/- 16,926) |
| test flat_bvh::bench::bench_build_120k_triangles_flat_bvh                |  97,173,840 ns/iter (+/- 6,132,352)    |  93,301,800 ns/iter (+/- 3,916,466) |
| test flat_bvh::bench::bench_build_12k_triangles_flat_bvh                 |   9,569,305 ns/iter (+/- 447,199)      |   9,162,365 ns/iter (+/- 242,592) |
| test flat_bvh::bench::bench_flatten_120k_triangles_bvh                   |   7,332,465 ns/iter (+/- 984,515)      |   7,517,620 ns/iter (+/- 537,332) |
| test flat_bvh::bench::bench_intersect_1200_triangles_flat_bvh            |         147 ns/iter (+/- 2)            |         179 ns/iter (+/- 2) |
| test flat_bvh::bench::bench_intersect_120k_triangles_flat_bvh            |         890 ns/iter (+/- 31)           |       1,121 ns/iter (+/- 21) |
| test flat_bvh::bench::bench_intersect_12k_triangles_flat_bvh             |         364 ns/iter (+/- 3)            |         473 ns/iter (+/- 6) |
| test ray::bench::bench_intersects_aabb                                   |       2,897 ns/iter (+/- 46)           |       4,475 ns/iter (+/- 23) |
| test ray::bench::bench_intersects_aabb_branchless                        |       3,469 ns/iter (+/- 40)           |               - |
| test ray::bench::bench_intersects_aabb_naive                             |       6,037 ns/iter (+/- 84)           |               - |
| test testbase::bench_intersect_120k_triangles_list                       |           7 ns/iter (+/- 0)            |           7 ns/iter (+/- 0) |
| test testbase::bench_intersect_120k_triangles_list_aabb                  |           7 ns/iter (+/- 0)            |           7 ns/iter (+/- 0) |
| test testbase::bench_intersect_sponza_list                               |           7 ns/iter (+/- 0)            |           7 ns/iter (+/- 0) |
| test testbase::bench_intersect_sponza_list_aabb                          |           7 ns/iter (+/- 0)            |           7 ns/iter (+/- 0) |

### CODE REVIEW

**Comment by marstaik** [comment]
Added the f32x3 optimization, It can be done quite easily to also handle f32x2,3,4, and f64x2,3,4, which I will do later.

Some new bench numbers:
| test | nalgebra |
|-----|------------|
| test bvh::bvh_impl::bench::bench_build_1200_triangles_bvh                |     777,386 ns/iter (+/- 22,640) |
| test bvh::bvh_impl::bench::bench_build_120k_triangles_bvh                |  99,323,460 ns/iter (+/- 1,955,579) |
| test bvh::bvh_impl::bench::bench_build_12k_triangles_bvh                 |   8,601,280 ns/iter (+/- 160,429) |
| test bvh::bvh_impl::bench::bench_build_sponza_bvh                        |  79,983,790 ns/iter (+/- 1,201,449) |
| test bvh::bvh_impl::bench::bench_intersect_1200_triangles_bvh            |         133 ns/iter (+/- 1) |
| test bvh::bvh_impl::bench::bench_intersect_120k_triangles_bvh            |         802 ns/iter (+/- 7) |
| test bvh::bvh_impl::bench::bench_intersect_12k_triangles_bvh             |         335 ns/iter (+/- 5) |
| test bvh::bvh_impl::bench::bench_intersect_sponza_bvh                    |       1,403 ns/iter (+/- 24) |
| test bvh::iter::bench::bench_intersect_128rays_sponza_iter               |     156,681 ns/iter (+/- 3,499) |
| test bvh::iter::bench::bench_intersect_128rays_sponza_vec                |     179,903 ns/iter (+/- 2,719) |
| test bvh::optimization::bench::bench_intersect_120k_after_optimize_00p   |         802 ns/iter (+/- 8) |
| test bvh::optimization::bench::bench_intersect_120k_after_optimize_01p   |     134,032 ns/iter (+/- 10,309) |
| test bvh::optimization::bench::bench_intersect_120k_after_optimize_10p   |   1,606,707 ns/iter (+/- 389,896) |
| test bvh::optimization::bench::bench_intersect_120k_after_optimize_50p   |   2,200,720 ns/iter (+/- 453,999) |
| test bvh::optimization::bench::bench_intersect_120k_with_rebuild_00p     |         801 ns/iter (+/- 8) |
| test bvh::optimization::bench::bench_intersect_120k_with_rebuild_01p     |         865 ns/iter (+/- 8) |
| test bvh::optimization::bench::bench_intersect_120k_with_rebuild_10p     |       1,887 ns/iter (+/- 26) |
| test bvh::optimization::bench::bench_intersect_120k_with_rebuild_50p     |       2,157 ns/iter (+/- 29) |
| test bvh::optimization::bench::bench_intersect_sponza_after_optimize_00p |       1,297 ns/iter (+/- 21) |
| test bvh::optimization::bench::bench_intersect_sponza_after_optimize_01p |       2,655 ns/iter (+/- 56) |
| test bvh::optimization::bench::bench_intersect_sponza_after_optimize_10p |       3,847 ns/iter (+/- 114) |
| test bvh::optimization::bench::bench_intersect_sponza_after_optimize_50p |       5,992 ns/iter (+/- 426) |
| test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_00p   |       1,300 ns/iter (+/- 20) |
| test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_01p   |       1,437 ns/iter (+/- 33) |
| test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_10p   |       1,814 ns/iter (+/- 29) |
| test bvh::optimization::bench::bench_intersect_sponza_with_rebuild_50p   |       2,370 ns/iter (+/- 48) |
| test bvh::optimization::bench::bench_optimize_bvh_120k_00p               |   1,158,720 ns/iter (+/- 7,658) |
| test bvh::optimization::bench::bench_optimize_bvh_120k_01p               |   2,232,760 ns/iter (+/- 89,304) |
| test bvh::optimization::bench::bench_optimize_bvh_120k_10p               |  12,362,530 ns/iter (+/- 1,015,947) |
| test bvh::optimization::bench::bench_optimize_bvh_120k_50p               |  58,228,630 ns/iter (+/- 11,172,616) |
| test bvh::optimization::bench::bench_randomize_120k_50p                  |   5,814,270 ns/iter (+/- 232,303) |
| test flat_bvh::bench::bench_build_1200_triangles_flat_bvh                |     790,160 ns/iter (+/- 13,313) |
| test flat_bvh::bench::bench_build_120k_triangles_flat_bvh                | 104,565,130 ns/iter (+/- 4,193,432) |
| test flat_bvh::bench::bench_build_12k_triangles_flat_bvh                 |  10,120,240 ns/iter (+/- 224,187) |
| test flat_bvh::bench::bench_flatten_120k_triangles_bvh                   |   6,722,320 ns/iter (+/- 1,746,270) |
| test flat_bvh::bench::bench_intersect_1200_triangles_flat_bvh            |         131 ns/iter (+/- 1) |
| test flat_bvh::bench::bench_intersect_120k_triangles_flat_bvh            |         889 ns/iter (+/- 15) |
| test flat_bvh::bench::bench_intersect_12k_triangles_flat_bvh             |         354 ns/iter (+/- 4) |
| test ray::bench::bench_intersects_aabb                                   |       2,560 ns/iter (+/- 9) |
| test testbase::bench_intersect_120k_triangles_list                       |           7 ns/iter (+/- 0) |
| test testbase::bench_intersect_120k_triangles_list_aabb                  |           7 ns/iter (+/- 0) |
| test testbase::bench_intersect_sponza_list                               |           7 ns/iter (+/- 0) |
| test testbase::bench_intersect_sponza_list_aabb                          |           7 ns/iter (+/- 0) |

**Review by svenstaro — CHANGES_REQUESTED** [review]
Hey, great stuff. There appears to be some dead code that should be handled.

File: src/testbase.rs
Comment (svenstaro) [inline src/testbase.rs:18]: Is this TODO current?

File: src/ray.rs
Comment (svenstaro) [inline src/ray.rs:224]: More dead code.

File: src/ray.rs
Comment (svenstaro) [inline src/ray.rs:127]: Really digging all this custom SIMD but I'm wondering whether perhaps [packed_simd](https://github.com/rust-lang/packed_simd) might be the way forward?

File: src/lib.rs
Comment (svenstaro) [inline src/lib.rs:76]: These features turn this lib into a nightly-only lib, don't they? If so, that should be avoided.

File: src/aabb.rs
Comment (svenstaro) [inline src/aabb.rs:149]: Some dead code here.

**Review by svenstaro — CHANGES_REQUESTED** [review]

**Comment by svenstaro** [comment]
@marstaik Are you likely to make more bigger contributions? I could give you contributor access if you're interested to help with development and maintenance but I couldn't find your email or Matrix for further communication. If you're interested, it might be good to talk first.

File: src/ray.rs
Comment (marstaik) [reply src/ray.rs:127]: packed_simd says on the readme that std::simd is the way forward

File: src/lib.rs
Comment (marstaik) [reply src/lib.rs:76]: packed_simd and std::simd both need nightly... I wonder if we can detect nightly and turn it on or off accordingly

File: src/ray.rs
Comment (svenstaro) [reply src/ray.rs:127]: Hm true. I didn't know there's still no good stabilized SIMD stuff in Rust. What do you think instead about [wide](https://github.com/Lokathor/wide)?

Comment (marstaik) [reply src/ray.rs:127]: I'll look into it and maybe trying Simba (what nalgebra uses under the hood) again. But with Simba I had performance magnitudes worse that with the pure simd, and I'm unsure why

Comment (svenstaro) [reply src/ray.rs:127]: There's also [simdeez](https://crates.io/crates/simdeez) which might be worth checking out.

**Comment by marstaik** [comment]
@svenstaro I can use SIMD with pure `core::arch::x86_64`, as nothing else really works (simba, wide, and most other libraries don't give me all of the intrinsics needed for good performance, such as shuffle).

The issue I am having is that I still need the specialization feature from nightly to be safely able to handle all cases and dimensions. I am currently trying to find a way to do this without, but it seems very convoluted.

I have a generic ray_intersects_bvh for any Vector size, but I need to manually specialize for the lower dimension cases, as well as for when x86_64 is available.

**Comment by svenstaro** [comment]
Well realistically people will mostly be using this on x86_64. However, even if we make good performance nightly only for the time being, that should be put behind a feature flag so people can still choose to use this on stable.

**Comment by svenstaro** [comment]
I'm not sure it was clear earlier: If we have to use nightly for the speedy stuff anyway, then might as well use the most convenient solution to achieve that so you don't necessarily have to switch to `core`.

File: src/testbase.rs
Comment (marstaik) [reply src/testbase.rs:18]: Yeah, I plan to expand the unit tests later to handle 2d and 4d cases

**Comment by marstaik** [comment]
I ended up sticking with `core::arch` as it was just way easier to use the native intrinsics than every libraries attempt at a custom wrapper for SIMD...

I added a feature flag called "full_simd" that only works with nightly enabled, allowing for specialization of the fast vectors.

I would like to add some benches and tests for 2d and 4d as well.

File: Cargo.toml
Comment (svenstaro) [inline Cargo.toml:33]: Make sure to document this in the README.

**Comment by svenstaro** [comment]
Please make sure to change the CI so it runs twice: Once with and once without the `full_simd` feature. Also, why not just call it `simd`?

**Review by svenstaro — CHANGES_REQUESTED** [review]

File: src/testbase.rs
Comment (svenstaro) [reply src/testbase.rs:18]: Would you like to do that in this PR?

**Comment by marstaik** [comment]
Hey, I fixed the CI builds up a bit and added the SIMD builds via nightly only. The maintainer of the script for the rust environment disappeared and CI complained node was out of date, so I just overhauled it a bit.

Clippy was complaining about BVH and AABB naming conventions, so I also renamed BVH, AABB, BVHNode to Bvh, Aabb, BvhNode, and etc. I kind of like it, but if you don't want to do a breaking change name wise I can revert those changes, and we can ignore them perhaps on clippy.

I fixed a couple of other issues clippy complained about as well.

**Comment by marstaik** [comment]
Also, the README probably needs an update to reflect new data about rebuilding times versus building the tree from scratch again, because the rebuild numbers have drastically changed. That and the generics/testing of f32/f64 N dimensions would be nice to get done in a separate commit.

File: src/aabb.rs
Comment (svenstaro) [inline src/aabb.rs:17]: Let's merge these.

File: src/ray/intersect_default.rs
Comment (svenstaro) [inline src/ray/intersect_default.rs:31]: How about removing the feature check here and moving this function into `intersect_x86_64.rs`? Then you could just slap a feature check on top of the `mod` lines and I think it'd be a bit cleaner.

**Review by svenstaro — CHANGES_REQUESTED** [review]

**Comment by svenstaro** [comment]
I've also added a dummy `CHANGELOG.md` on master and I'd like to ask you to write a notice there and if you feel like it a short migration guide as well. We haven't had a changelog before and I feel this is as good an opportunity as any to start one at last.

File: src/ray/intersect_default.rs
Comment (marstaik) [reply src/ray/intersect_default.rs:31]: I was thinking I might add some explicit SIMD for ARM as well, which is why I left it in default. It is also the default implementation which I thought made sense for it to be here. Its unfortunate that the only difference between the two is the `default` keyword.

File: src/utils.rs
Comment (marstaik) [reply src/utils.rs:26]: the build in `max` and `min` functions by both nalgebra and std require `Ord`, which f32 and f64 do not satisfy

File: src/ray/intersect_default.rs
Comment (svenstaro) [reply src/ray/intersect_default.rs:31]: Sure and I think that's fine. I was just commenting about code order unless I'm missing something. I'm thinking we should rather have
```rust
#[cfg(not(feature = "simd"))]
mod default_impl;

#[cfg(all(feature = "simd", target_arch = "x86_64"))]
mod x86_64_impl;

#[cfg(all(feature = "simd", target_arch = "aarch64"))]
mod aarch64_impl;
```

as opposed to `cfg`-ing away the specific implementations per file.

Comment (marstaik) [reply src/ray/intersect_default.rs:31]: I believe the ones in the `default` file are the only ones that are per function (and there are two).

The x86_64 file is completely guarded like you posted above already.

File: src/utils.rs
Comment (svenstaro) [reply src/utils.rs:26]: Might be worth adding that info in a comment. :)

**Comment by marstaik** [comment]
I think that should be everything :)

File: README.md
Comment (svenstaro) [inline README.md:98]: There are still a few miss-capitalized occurrences like this in prose and docs. Could you do another pass?

Comment (marstaik) [reply README.md]: Woops, fixed! I didn't actually find any in the docs though.

**Comment by marstaik** [comment]
@svenstaro is this ready for merge?

**Comment by svenstaro** [comment]
Will take another look in a bit.

File: Cargo.toml
Comment (svenstaro) [inline Cargo.toml:8]: Would you like to add yourself as author?

Comment (marstaik) [reply Cargo.toml:8]: Ah, sure. Is everything else alright? If so I can just merge it myself when I add it in.

**Comment by svenstaro** [comment]
I'm doing one final check now. One thing I was wondering about: Should we kick out optimization? It never performed super well it did degrade the tree quickly. Should we perhaps just focus on super quick tree building using multithreading and SIMD?

File: src/flat_bvh.rs
Comment (svenstaro) [inline src/flat_bvh.rs:47]:
```suggestion
    /// Creates a flat node from a `BVH` inner node and its `AABB`. Returns the next free index.
```

File: src/flat_bvh.rs
Comment (svenstaro) [inline src/flat_bvh.rs:468]:
```suggestion
    /// Benchmark the flattening of a BVH with 120,000 triangles.
```

File: src/ray/intersect_x86_64.rs
Comment (svenstaro) [inline src/ray/intersect_x86_64.rs:43]: Could you write a short blip of doc comment for each of these functions? I think it would make the file easier to scan.

File: src/testbase.rs
Comment (svenstaro) [reply src/testbase.rs:17]: @marstaik ping

File: src/testbase.rs
Comment (svenstaro) [inline src/testbase.rs:525]:
```suggestion
/// Benchmark intersecting the `triangles` list with `AABB` checks, but without acceleration
```

File: src/testbase.rs
Comment (svenstaro) [inline src/testbase.rs:535]:
```suggestion
            // First test whether the ray intersects the AABB of the triangle.
```

**Review by svenstaro — CHANGES_REQUESTED** [review]
Alright, very close now. I'm just missing some docs on the SIMD functions and some capitalization fixes.

**Comment by marstaik** [comment]
@svenstaro I'll make the requested changes and look into things later today - but before I do that I was wondering if the Aabb to AABB capitalization changes in the docs make sense - they are generally referring to them via their `class names` rather than the AABB as a concept.

It doesn't bother me either way but I wanted to note it.

**Comment by svenstaro** [comment]
That's a good point. Let's make them clickable where they refer to the specific classes by doing [`Aabb`].

**Comment by dbenson24** [comment]
Once this PR lands I'd be happy to take a look at adapting my work in #80. I think there's probably 3 PRs worth of stuff in that. The multithreaded builds, the replacement of optimize with the ability to add/remove individual nodes, and the additional shapes you can query with

**Comment by marstaik** [comment]
> I'm doing one final check now. One thing I was wondering about: Should we kick out optimization? It never performed super well it did degrade the tree quickly. Should we perhaps just focus on super quick tree building using multithreading and SIMD?

I think we need to re-look at this. The new rebuild times are really really fast (400%) faster eg 200 sec to 40 sec for 50% modified. It might actually be worth it now. Maybe we can look into a better self optimizing bvh on removals and additions as well.

We should definitely look at both and compare, especially after multithreaded building is done. I wonder if optimizing can be improved via multithreading too thoough. Maybe @dbenson24 can look into it, as my next focus will probably be better unit tests and some more docs.

**Comment by dbenson24** [comment]
If I recall correctly, the change to make optimize utilize the add/remove node functions was because that method produced trees with much faster traversal times compared to optimize

**Comment by marstaik** [comment]
I may have gone overboard on the document cleaning :)

File: Cargo.toml
Comment (svenstaro) [reply Cargo.toml:8]: True, you can do that.

**Comment by svenstaro** [comment]
Merging as-is. Amazing work!

**Pull request merged**
